### PR TITLE
Add float zero point support for 2-bit LUT GEMM in MatMulNBits

### PR DESCRIPTION
**Pull request by vraspar**
## Description

Adds support for float/float16 zero points in the 2-bit MatMulNBits LUT GEMM path, enabling AMD QAD/Quark 2-bit quantization which requires a fractional zero point of 1.5.

Addresses #28162

### Problem

QAD 2-bit quantization uses non-uniform levels `[-1, -1/3, 1/3, 1]`, expressed via `dequant = (q - 1.5) * scale`. The zero point 1.5 cannot be represented as a packed uint8 value. The existing LUT GEMM packing API only accepted `uint8_t*` zero points, and the fallback dequant path crashed with `ORT_ENFORCE(nbits_ == 4)` when encountering 2-bit + float ZP.

### Changes

**MLAS layer** — Widened `MlasLutGemmPack()` to accept `const void* QuantBZeroPoint` + `bool IsFloatZeroPoint`, following the existing `MlasQNBitGemmPackQuantBData` convention. The AVX2 packer reads float ZP values directly per quantization group when `IsFloatZeroPoint` is set, computing the same `(zp - midpoint) * scale` correction stored in the packed buffer. The compute kernel (`TMACComputeGemm_avx2`) is unchanged — it already consumes ZP as a float correction during accumulation.

**MatMulNBits CPU kernel** — Relaxed the PrePack early-exit guard to allow float ZP into the LUT GEMM path (not non-LUT paths). Added fp16→fp32 conversion for ZP tensors, matching how scales are already handled. Fixed the Compute() path to null out prepacked zero_points to avoid a null dereference in CheckInputs. Fixed the 2-bit fallback dequant path: relaxed the `nbits_==4` enforce, added inline 2-bit scalar dequant for float and MLFloat16 ZP with correct packed-B indexing for padded K shapes.

**Tests** — Added MLAS-level float ZP tests across block lengths 32/64/128 with ZP values {0, 1.5, 2, 3}. Added provider-level directed QAD tests (`zp=1.5`) verifying end-to-end correctness through the LUT GEMM path.

### Testing

- 72 MLAS LUT GEMM tests pass (including 36 new float ZP tests)
- 13 provider-level 2-bit tests pass (including new QAD float ZP tests)
- No regressions in existing uint8 ZP tests
- lintrunner clean

### Files changed

| File | Change |
|------|--------|
| `core/mlas/inc/mlas_qnbit.h` | API: `void*` ZP + `IsFloatZeroPoint` flag |
| `core/mlas/lib/qlutgemm.h` | Dispatch typedef update |
| `core/mlas/lib/qlutgemm.cpp` | Pass-through plumbing |
| `core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp` | Float ZP packing branch |
| `contrib_ops/cpu/quantization/matmul_nbits.cc` | PrePack guard, fallback fix, ZP validation |
| `test/mlas/unittest/test_sqlutgemm.cpp` | Float ZP MLAS tests |
| `test/mlas/bench/bench_lutgemm.cpp` | Updated call signature |
| `test/contrib_ops/matmul_2bits_test.cc` | Float ZP provider tests |

### CODE REVIEW

**Review by vraspar — COMMENTED** [review]
## Review: Add float zero point support for 2-bit LUT GEMM in MatMulNBits

Overall: Clean, well-scoped PR. The math is correct, the API extension is pragmatic, and test coverage is solid. Two items worth addressing:

---

### 1. Test reference uses wrong packed stride (both test files)

**`matmul_2bits_test.cc` -- `RunTest2BitsFloatZP`:**
`cpp
int64_t packed_idx = n * (K / 4) + k / 4;  // assumes K == packed_k
`

**`test_sqlutgemm.cpp` -- `TestFloatZeroPoint`:**
`cpp
size_t packed_idx = n * (K / (8 / BlkBitWidth)) + k / (8 / BlkBitWidth);  // same issue
`

`MlasQuantizeBlockwise` pads to `packed_k = k_blocks * block_size`. The kernel correctly uses the padded stride (`packed_k / 4`), and the fallback test (`Float32_2b_FloatZP_Fallback`) gets it right too. These two references diverge from the kernel when K is not a multiple of block_size. All current test cases happen to use aligned K, so it is not a live failure -- but it would produce wrong reference values if someone adds `RunTest2BitsFloatZP(1, 128, 100, 32, 1.5f)`. Suggest using the same `packed_k`-based stride and adding one non-aligned-K test case to prevent regression.

### 2. Duplicated ZP conversion in PrePack -- extract helper

The ~25-line fp16-to-fp32 ZP conversion + validation block is copy-pasted in two PrePack code paths (`input_idx == B` and `input_idx == scales` under `prefer_lut_gemm_`). If bfloat16 ZP or other changes come later, both copies need updating. Consider extracting into a small helper method or local lambda.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc [hunk 1]
```diff
@@ -870,23 +917,46 @@ Status MatMulNBits<float>::ComputeBUnpacked(const Tensor* a,
...
     // Hitting any of the below is very rare
     ORT_ENFORCE(column_wise_quant_, "Row-wise quantization is not supported for now");
-    ORT_ENFORCE(nbits_ == 4,
-                "Only 4b quantization is supported for unpacked compute using "
+    ORT_ENFORCE(nbits_ == 2 || nbits_ == 4,
+                "Only 2b and 4b quantization is supported for unpacked compute using "
                 "non-MLAS de-quantization for now");
 
     // !!!!!!!!!!!!!! naive implementation, need to be optimized !!!!!!!!!!!!!!
     if (zero_points && zero_points->IsDataType<float>()) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc:937]: This only handles `float` ZP in the fallback branch. If a model provides `MLFloat16` zero points with float activations (and the LUT GEMM session option is not set), execution falls through to the final `else` which casts the MLFloat16 buffer as `uint8_t*`, producing silent data corruption.

Suggested fix: add an `else if (zero_points && zero_points->IsDataType<MLFloat16>())` branch that converts to float before dequant (similar to how `ConvertFloatZeroPointsForLutGemm` handles it), or use `has_unquantized_zero_point_` to route both float and fp16 ZPs into this section.

The symmetric issue exists in `ComputeBUnpacked<MLFloat16>` (float ZP with fp16 activation).

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc [hunk 2]
```diff
@@ -870,23 +917,46 @@ Status MatMulNBits<float>::ComputeBUnpacked(const Tensor* a,
...
+      if (nbits_ == 2) {
+        ORT_ENFORCE(reorder_idx_data == nullptr,
+                    "g_idx (reorder index) is not supported for 2-bit quantization with float zero points");
+        // Simple 2-bit dequantization with float zero points
+        const float* float_zp = static_cast<const float*>(zero_points_data);
+        int32_t k_blocks = (static_cast<int32_t>(K_) + static_cast<int32_t>(block_size_) - 1) /
+                           static_cast<int32_t>(block_size_);
+        int32_t packed_k = k_blocks * static_cast<int32_t>(block_size_);
+        int32_t bytes_per_col = packed_k / 4;
+        for (int32_t n = 0; n < static_cast<int32_t>(N_); n++) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc]: Nit: `n * bytes_per_col + k / 4` uses `int32_t`. For consistency with the MLAS-level code which uses `size_t` for the same computation, and to future-proof against very large models, consider using `size_t` for `packed_idx` and `bytes_per_col`.

**Review by tianleiwu — COMMENTED** [review]
## Review Summary

The approach is sound: widening the MLAS API to accept `void*` ZP with a type flag, scalar fallback for 2-bit float ZP, and properly nulling `zero_points` in `Compute()` when LUT GEMM has consumed them. The `ConvertFloatZeroPointsForLutGemm` helper and the `compute_zp_correction` lambda in the AVX2 kernel are clean abstractions.

However, the cross-type ZP issue from the previous review round remains unaddressed: in `ComputeBUnpacked<float>`, an MLFloat16 zero-point tensor enters the uint8 cast path, producing garbage. Additionally, there's still no test coverage for the MLFloat16 ZP paths that this PR adds code for.

Overall quality is good — well-structured, good test coverage of the primary path, and clear commit messages explaining the progression of fixes.

File: onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp [hunk 3]
```diff
@@ -962,9 +963,24 @@ PackScalesAndZeroPoints_avx2_impl(
...
 
+    // Cast ZP pointers based on type
+    const uint8_t* uint8_zp = IsFloatZeroPoint ? nullptr : static_cast<const uint8_t*>(QuantBZeroPoint);
+    const float* float_zp = IsFloatZeroPoint ? static_cast<const float*>(QuantBZeroPoint) : nullptr;
+
     const size_t TotalBlocks = N * row_blks;
     ptrdiff_t MaxThreads = MlasGetMaximumThreadCount(ThreadPool);
 
+    // Lambda to compute ZP correction for a given (row, block) pair
+    auto compute_zp_correction = [&](size_t im, size_t blk_in_col, float scale) -> float {
```
Comment (tianleiwu) [inline onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp:978]: Minor optimization opportunity: `IsFloatZeroPoint` is a runtime value checked inside a potentially tight parallel loop. Since `HasZeroPoint` is already a template parameter, consider making `IsFloatZeroPoint` a template parameter too (adding 2 more instantiations) to enable compile-time branch elimination. This is low priority since packing is a one-time cost.

File: onnxruntime/test/contrib_ops/matmul_2bits_test.cc [hunk 4]
```diff
@@ -405,6 +405,254 @@ TEST(MatMulNBitsLutGemm, Float32_2Bits_Asymmetric_Batch32_256x256) {
...
+
+  SessionOptions so;
+  ASSERT_STATUS_OK(so.config_options.AddConfigEntry(kOrtSessionOptionsMlasLutGemm, "1"));
+
+  test.Config(so)
+      .ConfigEp(DefaultCpuExecutionProvider())
+      .RunWithConfig();
+}
+
+TEST(MatMul2Bits, Float32_2b_FloatZP_QAD) {
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/matmul_2bits_test.cc:490]: All new provider-level tests use `float` zero-point tensors. The implementation also adds `MLFloat16` ZP handling in `ConvertFloatZeroPointsForLutGemm` and in `ComputeBUnpacked<MLFloat16>`, but neither path has test coverage. Consider adding at least one directed test with `MLFloat16` zero points to verify the fp16 conversion path doesn't produce wrong results.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc [hunk 5]
```diff
@@ -190,13 +190,40 @@ class MatMulNBits final : public OpKernel {
                            const MatMulComputeHelper& helper) const;
 };
 
+// Helper to convert float/fp16 zero points to float32 for LUT GEMM prepack.
+// Returns pointer to float32 ZP data in the provided buffer.
+static const float* ConvertFloatZeroPointsForLutGemm(
+    const Tensor* zero_points,
+    size_t N, size_t K, size_t block_size,
+    std::vector<float>& zp_fp32_buf) {
+  size_t k_blocks = (K + block_size - 1) / block_size;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc:202]: **Suggestion — Stride consistency with the AVX2 packer**

This uses ceiling division: `k_blocks = (K + block_size - 1) / block_size`. However, the AVX2 packer (`PackScalesAndZeroPoints_avx2_impl`) indexes the float ZP array with stride `nb1 = K / BlkLen` (floor division). When `K % BlkLen != 0` (e.g., K=96, BlkLen=64 — which passes `MlasIsLutGemmAvailable`'s `K % 32 == 0` check), column indexing `float_zp[im * nb1 + blk_in_col]` reads incorrect values for columns > 0.

This is technically a pre-existing issue affecting scales indexing too, but the new float ZP path inherits it. Consider adding a defensive guard:
```cpp
ORT_ENFORCE(K % block_size == 0,
            "LUT GEMM float ZP packer requires K to be a multiple of block_size");
```
Or align this calculation with the packer by using `K / block_size` instead of ceiling.

File: onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp [hunk 6]
```diff
@@ -962,9 +963,24 @@ PackScalesAndZeroPoints_avx2_impl(
     const int midpoint = 1 << (bits - 1);  // 2 for 2-bit
     const uint8_t bits_mask = static_cast<uint8_t>((1 << bits) - 1);
 
+    // Cast ZP pointers based on type
+    const uint8_t* uint8_zp = IsFloatZeroPoint ? nullptr : static_cast<const uint8_t*>(QuantBZeroPoint);
+    const float* float_zp = IsFloatZeroPoint ? static_cast<const float*>(QuantBZeroPoint) : nullptr;
```
Comment (tianleiwu) [inline onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp]: **Nitpick — Template the branch for zero-cost dispatch**

The `IsFloatZeroPoint` check inside `compute_zp_correction` is evaluated per-block. Since `HasZeroPoint` is already a template parameter, `IsFloatZeroPoint` could be templated too (`PackScalesAndZeroPoints_avx2_impl<HasZeroPoint, IsFloatZeroPoint>`) to eliminate the runtime branch entirely. Low priority since this is packing code and the branch predictor handles constant-direction branches well.

**Review by tianleiwu — COMMENTED** [review]
More inline comments...

**Review by vraspar — COMMENTED** [review]
## Addressing review feedback (commit 004beb1e3d)

Pushed fixes for all outstanding review items. Summary:

### Fixes applied

| Change | Addresses |
|--------|-----------|
| **`K % BlkLen == 0` guard** in `MlasIsLutGemmAvailable` | Thread by @tianleiwu re stride consistency — the AVX2 packer uses floor division (`K / BlkLen`) which diverges from caller's ceiling division when K % BlkLen != 0. This is pre-existing (affects uint8 ZP too) but newly exposed. Guard is the safe minimal fix; full arbitrary-K support is a separate MLAS effort. |
| **Dynamic ZP guard** in PrePack | Own finding — if ZP is a non-constant input, `has_zp_input_` is false, LUT packs without ZP, and Compute() nulls ZP → silent wrong results. Added `has_zp_arg_` member + early return, similar to KleidiAI's dynamic ZP fallback (lines 377-386). |
| **`int32_t` → `size_t`** in fallback dequant loops | Thread 18 by @tianleiwu — consistent with MLAS conventions, prevents overflow for very large models. |
| **T3 constraint comments** at ZP dispatch points | Thread 17 by @tianleiwu — the cross-type ZP scenario (MLFloat16 ZP with float activations) cannot happen because the kernel registration constrains T3 to `{uint8_t, T1}`. Added comments at the dispatch points explaining this so future readers don't need to trace through kernel binding. |
| **MLFloat16 ZP fallback test** | Thread 20 by @tianleiwu — new test with MLFloat16 activations + MLFloat16 zero points exercising `ComputeBUnpacked<MLFloat16>` 2-bit dequant path. |

### Items deferred (with rationale)

- **Full arbitrary-K LUT GEMM support**: Requires fixing packer, sizing, and compute kernel stride math across MLAS. The `K % BlkLen` guard prevents the bug; no known model needs non-aligned K with LUT GEMM.
- **MLFloat16 LUT GEMM support**: `MatMulNBits<MLFloat16>::PrePack` exits early for unquantized ZP. Adding LUT support for fp16 activations is a feature request with no current user demand; the fallback dequant path handles it correctly.

File: onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp (hunk 3 above)
Comment (vraspar) [reply onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp:978]: Done in 004beb1e3d — `IsFloatZeroPoint` is now a template parameter of `PackScalesAndZeroPoints_avx2_impl` alongside `HasZeroPoint`. Uses `if constexpr` for compile-time branch elimination. 3 instantiations: `<true,true>`, `<true,false>`, `<false,false>`.

File: onnxruntime/test/contrib_ops/matmul_2bits_test.cc (hunk 4 above)
Comment (vraspar) [reply onnxruntime/test/contrib_ops/matmul_2bits_test.cc:490]: Added in 004beb1e3d — new test `MLFloat16_2b_MLFloat16ZP_Fallback` with MLFloat16 activations + MLFloat16 zero points (zp=1.5) exercising the `ComputeBUnpacked<MLFloat16>` 2-bit fallback dequant path. No LUT session option so it hits the scalar path directly.

File: onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp (hunk 6 above)
Comment (vraspar) [reply onnxruntime/core/mlas/lib/sqnbitgemm_lut_kernel_avx2.cpp]: Done in 004beb1e3d — same commit as thread 19. `IsFloatZeroPoint` is now a compile-time template parameter with `if constexpr` for zero-cost dispatch.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc (hunk 2 above)
Comment (vraspar) [reply onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc]: Fixed in 004beb1e3d — changed all `int32_t` to `size_t` in both the float and MLFloat16 2-bit fallback dequant loops. Loop variables, `packed_idx`, `bytes_per_col`, `k_blocks`, etc. are all `size_t` now, consistent with MLAS conventions.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc (hunk 1 above)
Comment (vraspar) [reply onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc:937]: Addressed in 004beb1e3d — added a comment at the dispatch point explaining why this can't happen: the kernel registration constrains T3 to `{uint8_t, T1}`, so for `MatMulNBits<float>`, ZP can only be `uint8_t` or `float`. MLFloat16 ZP with float activations is rejected at kernel binding time.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc (hunk 5 above)
Comment (vraspar) [reply onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc:202]: Addressed in 004beb1e3d — added `K % BlkLen == 0` guard to `MlasIsLutGemmAvailable()`. This prevents the floor-vs-ceiling stride divergence at the availability check level. Full arbitrary-K support (fixing packer + sizing math) is deferred as a separate MLAS effort since no known model needs non-aligned K with LUT GEMM and the guard prevents the bug entirely for ORT callers.

File: onnxruntime/test/contrib_ops/matmul_2bits_test.cc [hunk 7]
```diff
@@ -405,6 +405,333 @@ TEST(MatMulNBitsLutGemm, Float32_2Bits_Asymmetric_Batch32_256x256) {
...
+  // Quantize symmetrically
+  MlasQuantizeBlockwise<float, QBits>(
+      input1_vals.data(), scales.data(), nullptr,
+      input1_fp32_vals.data(), static_cast<int32_t>(block_size),
+      true, static_cast<int32_t>(K), static_cast<int32_t>(N),
+      static_cast<int32_t>(N), tp);
+
+  // Create float ZP: one per quantization group
+  int64_t k_blocks = (K + block_size - 1) / block_size;
+  std::vector<float> float_zp(N * k_blocks, zp_value);
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/matmul_2bits_test.cc:439]: Suggestion: this fills every float zero point with the same scalar value, so the LUT path can pass even if the new packer transposes or strides the `(N, k_blocks)` ZP layout incorrectly. Please add at least one case with a per-column/per-block pattern, and compute the reference from `float_zp[n * k_blocks + blk]`, so layout bugs in `PackScalesAndZeroPoints` are observable.

File: onnxruntime/test/contrib_ops/matmul_2bits_test.cc [hunk 8]
```diff
@@ -405,6 +405,333 @@ TEST(MatMulNBitsLutGemm, Float32_2Bits_Asymmetric_Batch32_256x256) {
...
+  test.AddAttribute<int64_t>("K", K);
+  test.AddAttribute<int64_t>("N", N);
+  test.AddAttribute<int64_t>("block_size", block_size);
+  test.AddAttribute<int64_t>("bits", QBits);
+  test.AddAttribute<int64_t>("accuracy_level", static_cast<int64_t>(0));
+
+  test.AddInput<float>("A", {M, K}, input0_fp32_vals, false);
+  test.AddInput<uint8_t>("B", {q_cols, k_blocks, q_rows / k_blocks}, input1_vals, true);
+  test.AddInput<float>("scales", {N, k_blocks}, scales, true);
+  test.AddInput<float>("zero_points", {N, k_blocks}, float_zp, true);
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/matmul_2bits_test.cc:476]: Suggestion: all of the added float-ZP provider tests still make `zero_points` an initializer. The new `prefer_lut_gemm_ && has_zp_arg_ && !has_zp_input_` guard is what prevents the LUT prepack path from ignoring runtime zero points, so it would be good to cover it with a LUT-compatible test that enables `kOrtSessionOptionsMlasLutGemm` but passes `zero_points` as non-constant (`is_initializer=false`).

**Review by tianleiwu — COMMENTED** [review]
Thanks for the follow-up updates. I rechecked the current head and did not find a new blocking correctness issue. I left two suggestion-level test coverage notes that would make the float zero-point changes harder to regress.

**Review by tianleiwu — APPROVED** [review]
## Review Summary

The implementation is correct and well-structured. No correctness, security, or performance issues found.

**Highlights:**

- **MLAS API widening** — `const void* QuantBZeroPoint` + `bool IsFloatZeroPoint` is clean type erasure. All callers updated consistently.
- **K % BlkLen guard** — The new `MlasIsLutGemmAvailable()` check that rejects `K % BlkLen != 0` closes a subtle floor/ceiling division mismatch between the caller and the AVX2 packer kernel.
- **PrePack guard logic** — The split into three distinct checks (`has_g_idx_`, `has_unquantized_zero_point_ && !prefer_lut_gemm_`, `prefer_lut_gemm_ && has_zp_arg_ && !has_zp_input_`) is correct. The dynamic ZP guard prevents silent zero-point ignorance when ZP is not a constant initializer.
- **AVX2 compile-time dispatch** — `<bool HasZeroPoint, bool IsFloatZeroPoint>` template with `if constexpr` ensures zero runtime overhead in the tight packing loop.
- **Compute path safety** — Nulling `zero_points` when `prefer_lut_gemm_ && packed_b_` prevents null dereference in `CheckInputs`. When the dynamic ZP guard fires (no prepack), `packed_b_` stays null and ZP is correctly fetched from context for the scalar fallback.
- **Test coverage** — Comprehensive: uniform ZP (QAD zp=1.5), varying per-block ZP (detects stride/transpose bugs), dynamic ZP fallback, non-aligned K, MLFloat16 ZP, MLAS-level tests across block lengths 32/64/128.

All four previously unresolved review threads have been addressed by the current head and resolved:
- Cross-typed ZP concern → correct per kernel registration constraints (T3 = {uint8_t, T1})
- Uniform ZP test concern → addressed by `VaryingPerBlock` test
- Dynamic ZP test concern → addressed by `DynamicZP` test with `is_initializer=false`
- MLFloat16 test concern → addressed by `MLFloat16_2b_MLFloat16ZP_Fallback` test
